public/app/features/dashboard/services/DashboardSrv.ts

```diff
@@ -152,17 +157,24 @@ export class DashboardSrv {
     return this.dashboard;
   }
 
-  save(clone: any, options: { overwrite?: any; folderId?: any }) {
-    options = options || {};
-    options.folderId = options.folderId >= 0 ? options.folderId : this.dashboard.meta.folderId || clone.folderId;
+  save(clone: any, { folderId, overwrite = false, message }: DashboardSaveOptions = {}) {
+    folderId = folderId >= 0 ? folderId : this.dashboard.meta.folderId || clone.folderId;
 
     return this.backendSrv
-      .saveDashboard(clone, options)
-      .then(this.postSave.bind(this, clone))
-      .catch(this.handleSaveDashboardError.bind(this, clone, options));
+      .saveDashboard(clone, { folderId, overwrite, message })
+      .then((data: any) => this.postSave(data))
+      .catch(this.handleSaveDashboardError.bind(this, clone, { folderId }));
   }
 
-  saveDashboard(options?: { overwrite?: any; folderId?: any; makeEditable?: any }, clone?: DashboardModel) {
+  saveDashboard(
+    clone?: DashboardModel,
+    {
+      makeEditable = false,
+      folderId,
+      overwrite = false,
+      message,
+    }: DashboardSaveOptions & { makeEditable?: boolean } = {}
+  ) {
```

Comment:
this function definition feels a bit messy and hard to read. Wonder if there is an easier way to wite this.

Reply:
I'll see what I can do

Reply:
> This is looking good now. Feel like the complexity is a bit lower now that this is handled in the dashboard init phase instead of in the UI component.
> 
>     * When I am using Test data datasource (gdev-testdata) I get no option to return with changes

@torkelo If the dashboard the panel is on is provisioned or otherwise can't be saved then the option to return with changes is disabled. Can undo if you'd prefer though.

Reply:
ah, that explains it. I think the provisioned check is not needed as you can edit provisioned dashboards in UI and then when you try to save you will see the json